llvm/lib/Target/Xtensa/XtensaBRegFixupPass.cpp

```diff
@@ -0,0 +1,188 @@
+#include "llvm/Target/TargetMachine.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "xtensa-breg-fixup-pass"
+
+class XtensaBRegFixup : public MachineFunctionPass {
+public:
+  static char ID;
+  XtensaBRegFixup() : MachineFunctionPass(ID) {}
+
+  const XtensaSubtarget *Subtarget;
+  static const XtensaInstrInfo *XtensaII;
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+  llvm::StringRef getPassName() const override {
+    return "Xtensa bool reg fixup pass";
+  }
+
+private:
+  bool VisitMBB(MachineBasicBlock &MBB);
+  bool VisitInstruction(const MachineBasicBlock::instr_iterator &MII);
+};
+
+char XtensaBRegFixup::ID = 0;
+const XtensaInstrInfo *XtensaBRegFixup::XtensaII;
+
+bool XtensaBRegFixup::runOnMachineFunction(MachineFunction &MF) {
+
+  Subtarget = &static_cast<const XtensaSubtarget &>(MF.getSubtarget());
+  XtensaII = static_cast<const XtensaInstrInfo *>(Subtarget->getInstrInfo());
+  bool Modified = false;
+
+  MachineFunction::iterator I = MF.begin(), E = MF.end();
+
+  LLVM_DEBUG(dbgs() << "********* " << getPassName() << " *********\n");
+
+  for (; I != E; ++I)
+    Modified |= VisitMBB(*I);
+  return Modified;
+}
+
+FunctionPass *llvm::createXtensaBRegFixupPass() {
+  return new XtensaBRegFixup();
+}
+
+bool XtensaBRegFixup::VisitMBB(MachineBasicBlock &MBB) {
+  bool Modified = false;
+  MachineBasicBlock::instr_iterator MII = MBB.instr_begin(),
+                                    E = MBB.instr_end();
+  MachineBasicBlock::instr_iterator NextMII;
+
+  // Iterate through the instructions in the basic block
+  for (; MII != E; MII = NextMII) {
+    NextMII = std::next(MII);
+    MachineInstr *MI = &*MII;
+
+    // Don't reduce bundled instructions or pseudo operations
+    if (MI->isBundle() || MI->isTransient())
+      continue;
+
+    Modified |= VisitInstruction(MII);
+  }
+
+  return Modified;
+}
+
+bool XtensaBRegFixup::VisitInstruction(
+    const MachineBasicBlock::instr_iterator &MII) {
+  MachineInstr *MI = &*MII;
+  MachineBasicBlock &MBB = *MI->getParent();
+  MachineFunction *MF = MBB.getParent();
+  const XtensaInstrInfo &TII =
+      *static_cast<const XtensaInstrInfo *>(MF->getSubtarget().getInstrInfo());
+  unsigned Opcode = MI->getOpcode();
+  unsigned RegBase = Xtensa::B0;
+
+  switch (Opcode) {
+  case Xtensa::MOVBA_P2: {
+    /*
+      MOVBA_P2 Breg, Dst1, Dst2, Src
+       |
+       V
+      RSR       Dst1, BREG
+      LoadImmediate Dst2, BregMask
+      AND       Dst2, Dst2, Dst1
+      SLLI      Dst1, Src, BregShift
+      OR        Dst2, Dst2, Dst1
+      WSR       BREG, Dst2
+      */
+    MachineOperand Breg = MI->getOperand(0);
+    MachineOperand Dst1 = MI->getOperand(1);
+    MachineOperand Dst2 = MI->getOperand(2);
+    MachineOperand Src = MI->getOperand(3);
+    DebugLoc DL = MI->getDebugLoc();
+    unsigned RegNo = Breg.getReg().id() - RegBase;
+    int64_t Mask = 0xffff & (~(1 << RegNo));
+
+    BuildMI(MBB, MI, DL, TII.get(Xtensa::RSR)).add(Dst1).addReg(Xtensa::BREG);
```

Comment:
Should be `MachineInstrBuilder MIB = ...`
Required by LLVM_DEBUG statement below.